gorm/transaction_test.go

```diff
@@ -329,6 +426,440 @@ func TestContext(t *testing.T) {
+				if err != nil {
+					t.Error("Received an error beginning transaction")
+				}
+				if txn3 == nil {
+					t.Error("Did not receive a transaction from context")
+				}
+				// Case: Transaction begin is idempotent
+				if txn1 != txn3 {
+					t.Error("Got a different txn than was opened before")
+				}
+				test.txOpts = &sql.TxOptions{}
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxTxnOptMismatch {
+					t.Error("begin transaction should fail with an error TxnOptionsMismatch")
+				}
+			} else {
+				txn1, err := beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != nil {
+					t.Error("Received an error beginning transaction")
+				}
+				if txn1 == nil {
+					t.Error("Did not receive a transaction from context")
+				}
+				if err := mock.ExpectationsWereMet(); err != nil {
+					t.Errorf("failed to begin transaction for read-write db - %s", err)
+				}
+				test.txOpts.ReadOnly = true
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxTxnOptMismatch {
+					t.Error("begin transaction should fail with an error TxOptionsMismatch")
+				}
+				test.txOpts.ReadOnly = false
+				test.txOpts.Isolation = sql.LevelSerializable
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxTxnOptMismatch {
+					t.Error("begin transaction should fail with an error TxOptionsMismatch")
+				}
+				test.txOpts.Isolation = sql.LevelDefault
+				test.withOpts = readOnly
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxDBOptMismatch {
+					t.Error("begin transaction should fail with an error DBOptionsMismatch")
+				}
+				test.withOpts = readWrite
+				txn3, err := beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != nil {
+					t.Error("Received an error beginning transaction")
+				}
+				if txn3 == nil {
+					t.Error("Did not receive a transaction from context")
+				}
+				// Case: Transaction begin is idempotent
+				if txn1 != txn3 {
+					t.Error("Got a different txn than was opened before")
+				}
+				test.txOpts.ReadOnly = true
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxTxnOptMismatch {
+					t.Error("begin transaction should fail with an error TxOptionsMismatch")
+				}
+				test.txOpts.ReadOnly = false
+				test.txOpts.Isolation = sql.LevelSerializable
+				_, err = beginFromContextWithOptions(ctx, test.withOpts, test.txOpts)
+				if err != ErrCtxTxnOptMismatch {
+					t.Error("begin transaction should fail with an error TxOptionsMismatch")
+				}
+				txn4, err := beginFromContextWithOptions(ctx, test.withOpts, nil)
+				if err != nil {
+					t.Error("Received an error beginning transaction")
+				}
+				if txn4 == nil {
+					t.Error("Did not receive a transaction from context")
+				}
+				// Case: Transaction begin is idempotent
+				if txn1 != txn4 {
```

Comment:
i think we previous checked that `txn1 != nil` so you could omit the `txn4 == nil` case if you wanted to, but it works either way and you might want the more precise error message for the nil case, so not blocking.